### PR TITLE
[Foundation] Add support to ignore the cookies in the NSUrlSessionHandler.

### PR DESCRIPTION
**Pull request by mandel-macaque**
This change allows to ignore the use of cookies and cookie containers in
the NSUrlSessionHandler. There are two different cookie containers to
consider:

1. The native NSHttoCookieStorage.
2. The managed CookieContainer.

If the native one is set to null, the native code will not use a cookie
storage, which is used as a flag to ignore the managed one.

There is an interesting situation, we allow different types of sessions.
From the cookie storage point of view, Default and Background sessions
are the same, but Ephemeral is not, since we only want to store in ram
the cookies and do not share them.

This supposes a problem because Apple does not provide any API that will
allow to determine the session type use in the configuration. The
workaround has been to hide the direct native call for the configuration
and add an enum value that can later be accessed in the
NSUrlSessionHandler. Of course things cannot be that easy. When a
session is created with the configuration, it creates a copy, and the
internal session configuration does not longer have the flag, therefore,
we need to store the session type in the handler.

Fixes: https://github.com/xamarin/xamarin-macios/issues/7659

### CODE REVIEW

File: src/Foundation/NSUrlSessionConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+			Background,
+			Ephemeral,
+		}
+
+		SessionConfigurationType configurationType = SessionConfigurationType.Default; 
+
+		public SessionConfigurationType SessiontType {
+			get => configurationType;
+			private set => configurationType = value;
+		}
```
Comment (chamons) [inline src/Foundation/NSUrlSessionConfiguration.cs]: I believe this is the same as:

```csharp
		public SessionConfigurationType SessiontType { get; private set; } = SessionConfigurationType.Default;
```

File: src/Foundation/NSUrlSessionConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	{
+		public enum SessionConfigurationType {
+			Default,
+			Background,
+			Ephemeral,
+		}
+
+		SessionConfigurationType configurationType = SessionConfigurationType.Default; 
+
+		public SessionConfigurationType SessiontType {
```
Comment (chamons) [inline src/Foundation/NSUrlSessionConfiguration.cs]: SessiontType -> SessionType

File: tests/monotouch-test/Foundation/NSUrlSessionConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using MonoTouch.ObjCRuntime;
+#endif
+
+namespace MonoTouchFixtures.Foundation {
+
+	[TestFixture]
+	public class NSUrlSessionConfigurationTest {
+
+		[Test]
+		public void TestSessionTypeDefatul ()
```
Comment (chamons) [inline tests/monotouch-test/Foundation/NSUrlSessionConfiguration.cs]:
```suggestion
		public void TestSessionTypeDefault ()
```

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 4]
```diff
@@ -198,6 +198,106 @@ public void TestNSurlSessionHandlerCookieContainerSetCookie ()
...
+
+					var nativeHandler = new NSUrlSessionHandler () {
+						AllowAutoRedirect = true,
+						UseCookies = false,
+					};
+					var nativeClient = new HttpClient (nativeHandler);
+					var nativeResponse = await nativeClient.GetAsync (url);
+					nativeSetCookieResult = await nativeResponse.Content.ReadAsStringAsync ();
+
+					// got the response, perofm a second queries to the cookies endpoint to get
```
Comment (chamons) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs]:
```suggestion
					// got the response, preform a second queries to the cookies endpoint to get
```

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 5]
```diff
@@ -198,6 +198,106 @@ public void TestNSurlSessionHandlerCookieContainerSetCookie ()
...
+			Assert.IsNotNull (nativeCookieResult, "Native cookies result");
+			Assert.IsFalse (nativeCookieResult.Contains ("chocolate-chip"));
+			var cookiesFromServer = cookieContainer.GetCookies (new Uri (url));
+			Assert.AreEqual (0, cookiesFromServer.Count, "Cookies received from server.");
+		}
+
+		[Test]
+		public void TestNSUrlSessionEphemeralDisabledCookies ()
+		{
+			// assert we do throw an exception with ephmeral configs.
```
Comment (chamons) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs]:
```suggestion
			// assert we do throw an exception with ephemeral configs.
```

File: src/Foundation/NSUrlSessionConfiguration.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Foundation/NSUrlSessionConfiguration.cs]: It is.

File: src/Foundation/NSUrlSessionConfiguration.cs (hunk 2 above)
Comment (mandel-macaque) [reply src/Foundation/NSUrlSessionConfiguration.cs]: me and my typos :/

File: src/Foundation/NSUrlSessionConfiguration.cs (hunk 1 above)
Comment (chamons) [reply src/Foundation/NSUrlSessionConfiguration.cs]: My point is that is a lot shorter :)

File: tests/monotouch-test/System.Net.Http/MessageHandlers.cs [hunk 6]
```diff
@@ -198,6 +198,106 @@ public void TestNSurlSessionHandlerCookieContainerSetCookie ()
...
+			Assert.IsFalse (nativeCookieResult.Contains ("chocolate-chip"));
+			var cookiesFromServer = cookieContainer.GetCookies (new Uri (url));
+			Assert.AreEqual (0, cookiesFromServer.Count, "Cookies received from server.");
+		}
+
+		[Test]
+		public void TestNSUrlSessionEphemeralDisabledCookies ()
+		{
+			// assert we do throw an exception with ephemeral configs.
+			var config = NSUrlSessionConfiguration.EphemeralSessionConfiguration;
```
Comment (prmallic) [inline tests/monotouch-test/System.Net.Http/MessageHandlers.cs]: `using` here?

Comment (mandel-macaque) [reply tests/monotouch-test/System.Net.Http/MessageHandlers.cs]: Good catch

**Comment by mandel-macaque** [comment]
Known issue: https://github.com/xamarin/maccore/issues/581

**Pull request merged**